src/Intervals.php

```diff
@@ -0,0 +1,369 @@
+        }
+
+        $constraints = $constraint->getConstraints();
+
+        $intervalGroups = array();
+        $devGroups = array();
+        foreach ($constraints as $c) {
+            $res = self::get($c);
+            $intervalGroups[] = $res['intervals'];
+            $devGroups[] = $res['devConstraints'];
+        }
+
+        $dev = array();
+        if ($constraint->isDisjunctive()) {
+            foreach ($devGroups as $group) {
+                foreach ($group as $c) {
+                    if (!isset($dev[(string) $c])) {
+                        $dev[(string) $c] = $c;
+                    }
+                }
+            }
+
+            foreach ($dev as $i => $c) {
+                if ($c->getOperator() === '!=') {
+                    foreach ($dev as $j => $c2) {
+                        if ($i === $j) {
+                            continue;
+                        }
+                        $op = $c2->getOperator();
+                        // != dev-foo || != dev-bar -> *
+                        if ($op === '!=' && $c->getVersion() !== $c2->getVersion()) {
+                            $dev = array();
+                            break 2;
+                        }
+                        // != dev-foo || == dev-foo -> *
+                        if ($op === '==' && $c->getVersion() === $c2->getVersion()) {
+                            $dev = array();
+                            break 2;
+                        }
+                        // != dev-foo || == dev-master -> != dev-foo
+                        if ($op === '==') {
+                            unset($dev[$j]);
+                            continue;
+                        }
+                    }
+                }
+            }
+        } else {
+            $blacklist = array();
+            foreach ($devGroups as $i => $group) {
+                foreach ($group as $j => $c) {
+                    // all != constraints are kept
+                    if ($c->getOperator() === '!=') {
+                        $dev[(string) $c] = $c;
+                        continue;
+                    }
+
+                    $otherGroupMatches = 0;
+                    foreach ($devGroups as $i2 => $group2) {
+                        if ($i2 === $i) {
+                            continue;
+                        }
+
+                        foreach ($group2 as $j2 => $c2) {
+                            if ((string) $c2 === (string) $c || $c2 === self::anyDev()) {
+                                $otherGroupMatches++;
+                            }
+
+                            // != x && == x cancel each other, make sure none of these appears in the output
+                            if ($c->getOperator() === '==' && $c2->getOperator() === '!=' && $c->getVersion() === $c2->getVersion()) {
+                                $blacklist[] = (string) $c;
```

Comment:
should probably also be a map rather than array, so you don't unset the same thing multiple times.